lightning-graph-sync/src/processing.rs

```diff
@@ -0,0 +1,565 @@
+use std::io;
+use std::io::Read;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use bitcoin::BlockHash;
+use bitcoin::secp256k1::PublicKey;
+
+use lightning::ln::msgs::{
+	DecodeError, ErrorAction, LightningError, OptionalField, UnsignedChannelUpdate,
+};
+use lightning::routing::network_graph;
+use lightning::util::ser::{BigSize, Readable};
+
+use crate::error::GraphSyncError;
+
+/// Backdating is used as a mechanism to avoid accidentally overriding higher-fidelity gossip
+/// updates. Current delta amount is set to a week flat. Note that gossip updates are purged at
+/// the two-week-mark.
+///
+/// In the future, we will probably start delaying the purging of unsigned updates to their
+/// four-week-marks, or perhaps simply trust a provided server-timestamp that marks
+/// data receipt, as opposed to originally signed value. TODO: TBD
+const TIMESTAMP_RETROGRESSION_DELTA: u32 = 7 * 24 * 3600;
+
+/// Update network graph from binary data
+///
+/// `network_graph`: network graph to be updated
+///
+/// `update_data`: `&[u8]` binary stream that comprises the update data
+pub fn update_network_graph(
+	network_graph: &network_graph::NetworkGraph,
+	update_data: &[u8],
+) -> Result<(), GraphSyncError> {
+	let mut read_cursor = io::Cursor::new(update_data);
+	read_network_graph(&network_graph, &mut read_cursor)
+}
+
+pub(crate) fn read_network_graph<R: Read>(
+	network_graph: &network_graph::NetworkGraph,
+	mut read_cursor: &mut R,
+) -> Result<(), GraphSyncError> {
+	let mut prefix = Vec::with_capacity(4);
+	prefix.resize(4, 0);
+	read_cursor.read_exact(&mut prefix)?;
+
+	let _deserialize_unsigned = match &prefix[..] {
+		&[76, 68, 75, 1] => true,
+		_ => {
+			return Err(GraphSyncError::ProcessingError(
+				"Prefix must equal 'LDK' and the version byte 1".to_string(),
+			));
+		}
+	};
+
+	// let chain_source: Option<Box<dyn lightning::chain::Access>> = None;
+	let current_timestamp = SystemTime::now()
+		.duration_since(UNIX_EPOCH)
+		.unwrap()
+		.as_secs() as u32;
+
+	let retrogressed_timestamp = current_timestamp - TIMESTAMP_RETROGRESSION_DELTA; // back-date updates by one week
+
+	let chain_hash: BlockHash = Readable::read(read_cursor)?;
+	let latest_seen_timestamp: u32 = Readable::read(read_cursor)?; //
+	let emulated_last_seen_timestamp = u32::max(latest_seen_timestamp, retrogressed_timestamp);
+
+	let node_id_count: u32 = Readable::read(read_cursor)?;
+	let mut node_ids: Vec<PublicKey> = Vec::with_capacity(node_id_count as usize);
```

Comment:
I believe a corrupt file can OOM us here by setting this value to 4G, causing us to allocate 4G*32 bytes. We should bound it by some sensible maximum memory usage (its okay if we have to reallocate once or twice).

Reply:
true. You figure I should just do a silent cpm::min or have a maximum value and return an error if it's exceeded?